File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
import queue


# 모든 경우 탐색이니까 dfs..! 재귀를 이용해보자 근데 재귀 한계를 넘어가면 어떻게 하징 ..
def solution(numbers:list, target:int):

    return dfs(numbers,0,0,target)


def dfs(numbers:list, current:int,index:int,target:int)->int:
    # 종료 조건
    if index== len(numbers):
        # 만약 나온 결과물이 타겟과 동일하면 카운트 +1
        if current == target:
            return 1
        else:
            return 0

    # 재귀인데
    # +재귀
    sum_p = dfs(numbers,current+numbers[index],index+1,target)
    # - 재귀
    sum_m = dfs(numbers,current-numbers[index],index+1,target)

    return sum_p+sum_m

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
        public int solution(int[] numbers, int target) {
            return recur(0,target,0,numbers.length,numbers);
         }
         
         private int recur(int i,int target,int temp,int stop,int[] nums){
            if(i==stop){
                if (temp==target){
                    return 1;
                }else return 0;
            }
            
            int p_sum=recur(i+1,target,temp+nums[i],stop,nums);
            int m_sum=recur(i+1,target,temp-nums[i],stop,nums);
            return p_sum+m_sum;
         }
}